File: library/src/main/java/com/permissionx/tianjiaodev/java/InvisibleFragment.java
package com.permissionx.tianjiaodev.java;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class InvisibleFragment extends Fragment {

    private IInvisibleListener mEachCallback;
    private FragmentActivity mActivity;

    private static final int PERMISSION_CODE = 1;
    private static final int SETTING_CODE = 2;

    private Map<String, Boolean> mPermission;

    public void requestEachPermissions(@NonNull Map<String, Boolean> permissions, IInvisibleListener callback) {
        mEachCallback = callback;
        mPermission = permissions;

        Set<String> keys = mPermission.keySet();
        String[] p = keys.toArray(new String[keys.size()]);
        requestPermissions(p, PERMISSION_CODE);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        mActivity = getActivity();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == PERMISSION_CODE) {
            List<String> deniedList = new ArrayList<>();
            for (int i = 0; i < grantResults.length; i++) {
                int grantResult = grantResults[i];
                String name = permissions[i];
                if (grantResult != PackageManager.PERMISSION_GRANTED) {
                    deniedList.add(name);
                    //如果权限是必要的
                    if (mPermission.get(name)) {
                        showTipDialog();
                    } else {
                        if (!ActivityCompat.shouldShowRequestPermissionRationale(mActivity, name)) {
                            //拒绝，不要再询问
                        }
                    }
                }
            }

            boolean allGranted = deniedList.isEmpty();
            mEachCallback.onGrant(allGranted, deniedList);
        }
    }

    private void showTipDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(mActivity);
        builder.setTitle("Tips");

        builder.setMessage("Authorization is required to enable the functionality.\nPlease go ahead and set up.");
        builder.setPositiveButton("Go", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
                settingToActivity();
            }
        });
        builder.setNegativeButton("Close", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
                mActivity.finish();
                System.exit(0);
            }
        });

        builder.show();
    }

    private void settingToActivity() {
        Intent intent = new Intent();
        intent.setAction("android.settings.APPLICATION_DETAILS_SETTINGS");
        intent.setData(Uri.parse("package:" + mActivity.getPackageName()));
        mActivity.startActivityForResult(intent, SETTING_CODE);
    }

}

File: library/src/main/java/com/permissionx/tianjiaodev/InvisibleFragment.kt
package com.permissionx.tianjiaodev

import android.content.pm.PackageManager
import androidx.fragment.app.Fragment

//typealias 可以用于给任意类型指定一个别名
//接收Boolean, List<String>两种类型的参数
typealias PermissionCallback = (Boolean, List<String>) -> Unit

class InvisibleFragment : Fragment() {

    private var callback: PermissionCallback? = null

    //vararg 可变长度的permission
    fun requestNow(cb: PermissionCallback, vararg permission: String) {
        callback = cb
        requestPermissions(permission, 1)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        if (requestCode == 1) {
            //用来记录所有被用户拒绝的权限
            val deniedList = ArrayList<String>()

            for ((index, result) in grantResults.withIndex()) {
                if (result != PackageManager.PERMISSION_GRANTED) {
                    deniedList.add(permissions[index])
                }
            }

            //是否所有申请的权限均已被授权的标识，deniedList为空就是均被授权
            val allGranted = deniedList.isEmpty()
            callback?.let { it(allGranted, deniedList) }
        }
    }
}